File: lib/server/models/Instrument.js
const mongoose = require('mongoose');

const instrumentSchema = new mongoose.Schema({
  name: {
    type: String,
    required: true,
    trim: true
  },

  category: {
    type: String,
    required: true,
    enum: ['Гитаралар', 'Пернетақталы', 'Ұрмалы', 'Үрмелі', 'Шекті', 'Бас']
  },

  brand: {
    type: String,
    required: true
  },

  model: {
    type: String,
    required: true
  },

  description: {
    type: String,
    required: true
  },

  pricePerHour: {
    type: Number,
    required: true,
    min: 0
  },

  pricePerDay: {
    type: Number,
    required: true,
    min: 0
  },

  imageUrls: [{
    type: String
  }],

  rating: {
    type: Number,
    default: 0,
    min: 0,
    max: 5
  },

  reviewsCount: {
    type: Number,
    default: 0
  },

  location: {
    type: String,
    required: true
  },

  condition: {
    type: String,
    enum: ['Керемет', 'Жақсы', 'Қанағаттанарлық'],
    default: 'Жақсы'
  },

  isAvailable: {
    type: Boolean,
    default: true
  },

  isPublished: {
    type: Boolean,
    default: true
  },

  moderation: {
    removedBy:      { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null },
    removalReason:  { type: String, default: null },
    removedAt:      { type: Date, default: null },
    appeal: {
      message:    { type: String, default: null },
      status:     { type: String, enum: ['none', 'pending', 'resolved'], default: 'none' },
      submittedAt:{ type: Date, default: null },
      resolvedAt: { type: Date, default: null },
      resolvedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null },
      resolution: { type: String, enum: ['none', 'restored', 'rejected'], default: 'none' },
    }
  },

  features: [{
    type: String
  }],

  ownerId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },

  ownerName: {
    type: String,
    required: true
  },

  createdAt: {
    type: Date,
    default: Date.now
  },

  updatedAt: {
    type: Date,
    default: Date.now
  }
});

instrumentSchema.index({ ownerId: 1, createdAt: -1 });
instrumentSchema.index({ category: 1, isAvailable: 1 });
instrumentSchema.index({ name: 'text', description: 'text' });

instrumentSchema.pre('save', function(next) {
  this.updatedAt = Date.now();
  next();
});

module.exports = mongoose.model('Instrument', instrumentSchema);

File: lib/models/instrument.dart
class Instrument {
  final String id;
  final String name;
  final String category;
  final String brand;
  final String model;
  final String description;
  final double pricePerHour;
  final double pricePerDay;
  final List<String> imageUrls;
  final double rating;
  final int reviewsCount;
  final String location;
  final String condition;
  final bool isAvailable;
  final List<String> features;
  final String ownerId;
  final String ownerName;
  final DateTime createdAt;

  Instrument({
    required this.id,
    required this.name,
    required this.category,
    required this.brand,
    required this.model,
    required this.description,
    required this.pricePerHour,
    required this.pricePerDay,
    required this.imageUrls,
    this.rating = 0.0,
    this.reviewsCount = 0,
    required this.location,
    this.condition = 'good',
    this.isAvailable = true,
    this.features = const [],
    required this.ownerId,
    required this.ownerName,
    required this.createdAt,
  });

  bool get hourlyAvailable => pricePerHour > 0;

  static const _base = 'https://rentmuss-production.up.railway.app';
  static String _absUrl(String url) =>
      (url.isEmpty || url.startsWith('http')) ? url : '$_base$url';

  factory Instrument.fromJson(Map<String, dynamic> json) {
    return Instrument(
      id: json['_id'] ?? json['id'] ?? '',
      name: json['name'] ?? '',
      category: json['category'] ?? '',
      brand: json['brand'] ?? '',
      model: json['model'] ?? '',
      description: json['description'] ?? '',
      pricePerHour: (json['pricePerHour'] ?? 0).toDouble(),
      pricePerDay: (json['pricePerDay'] ?? 0).toDouble(),
      imageUrls: json['imageUrls'] != null
          ? List<String>.from(json['imageUrls']).map(_absUrl).toList()
          : [],
      rating: (json['rating'] ?? 0).toDouble(),
      reviewsCount: json['reviewsCount'] ?? 0,
      location: json['location'] ?? '',
      condition: json['condition'] ?? 'good',
      isAvailable: json['isAvailable'] ?? true,
      features:
          json['features'] != null ? List<String>.from(json['features']) : [],
      ownerId: json['ownerId'] ?? '',
      ownerName: json['ownerName'] ?? '',
      createdAt:
          json['createdAt'] != null
              ? DateTime.parse(json['createdAt'])
              : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'category': category,
      'brand': brand,
      'model': model,
      'description': description,
      'pricePerHour': pricePerHour,
      'pricePerDay': pricePerDay,
      'imageUrls': imageUrls,
      'rating': rating,
      'reviewsCount': reviewsCount,
      'location': location,
      'condition': condition,
      'isAvailable': isAvailable,
      'features': features,
      'ownerId': ownerId,
      'ownerName': ownerName,
      'createdAt': createdAt.toIso8601String(),
    };
  }
}
